shared/src/commonMain/kotlin/io/customer/shared/tracking/queue/QueueRunner.kt

```diff
@@ -0,0 +1,101 @@
+package io.customer.shared.tracking.queue
+
+import io.customer.shared.database.TrackingTask
+import io.customer.shared.database.TrackingTaskQueryHelper
+import io.customer.shared.sdk.meta.Workspace
+import io.customer.shared.tracking.api.*
+import io.customer.shared.tracking.api.model.*
+import io.customer.shared.tracking.constant.QueueTaskStatus
+import io.customer.shared.tracking.model.Activity
+import io.customer.shared.tracking.model.Task
+import io.customer.shared.tracking.model.TaskResponse
+import io.customer.shared.util.JsonAdapter
+import io.customer.shared.util.Logger
+
+/**
+ * Class responsible for sending tasks to server by communicating with network layer and updating
+ * tasks status and responses in database.
+ */
+internal interface QueueRunner {
+    suspend fun runQueueForTasks(pendingTasks: List<TrackingTask>): Result<Boolean>
+}
+
+internal class QueueRunnerImpl(
+    private val logger: Logger,
+    private val jsonAdapter: JsonAdapter,
+    private val workspace: Workspace,
+    private val trackingTaskQueryHelper: TrackingTaskQueryHelper,
+    private val trackingHttpClient: TrackingHttpClient,
+) : QueueRunner {
+    override suspend fun runQueueForTasks(pendingTasks: List<TrackingTask>): Result<Boolean> {
+        return kotlin.runCatching {
+            trackingTaskQueryHelper.updateTasksStatus(
+                status = QueueTaskStatus.SENDING,
+                tasks = pendingTasks,
+            )
+            val response = processBatchTasks(pendingTasks = pendingTasks)
+            return@runCatching response != null && !response.isServerUnavailable
+        }.onFailure { ex ->
+            logger.error("Failed to run queue for ${pendingTasks.size} tasks with error: ${ex.message}")
+            trackingTaskQueryHelper.updateTasksStatus(
+                status = QueueTaskStatus.FAILED,
+                tasks = pendingTasks,
+            )
+        }
+    }
+
+    @Throws(Exception::class)
+    private suspend fun processBatchTasks(pendingTasks: List<TrackingTask>): BatchTrackingResponse? {
+        val activities = pendingTasks.map { task ->
+            jsonAdapter.fromJSON(Activity::class, task.activityJson)
+        }
+        val result = trackingHttpClient.track(
+            tasks = activities.mapIndexed { index, activity ->
+                Task(
+                    identityType = workspace.identityType,
+                    profileIdentifier = pendingTasks[index].identity,
+                    activity = activity,
+                )
+            },
```

Comment:
I think we might not be able to do that. Mainly because we are parsing json from `TrackingTask` into `Activity` just above this, so passing `TrackingTask` to `trackingHttpClient.track()` won't be possible as it doesn't really know about it. I'll think more on this though.

Also looking again to it, I see that we can merge multiple `map` calls in this method, will do that. But `TrackingTask` will still be mapped to `Task` before being passed to `trackingHttpClient.track()`.